File: src/main/kotlin/com/linqingying/cangjie/resolve/calls/context/BasicCallResolutionContext.java
package com.linqingying.cangjie.resolve.calls.context;


import com.linqingying.cangjie.config.LanguageVersionSettings;
import com.linqingying.cangjie.descriptors.BindingTrace;
import com.linqingying.cangjie.psi.Call;
import com.linqingying.cangjie.psi.CjExpression;
import com.linqingying.cangjie.resolve.StatementFilter;
import com.linqingying.cangjie.resolve.calls.components.InferenceSession;
import com.linqingying.cangjie.resolve.calls.model.MutableDataFlowInfoForArguments;
import com.linqingying.cangjie.resolve.calls.smartcasts.DataFlowInfo;
import com.linqingying.cangjie.resolve.calls.smartcasts.DataFlowValueFactory;
import com.linqingying.cangjie.resolve.scopes.LexicalScope;
import com.linqingying.cangjie.types.CangJieType;
import com.linqingying.cangjie.types.expressions.ContextConfig;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class BasicCallResolutionContext extends CallResolutionContext<BasicCallResolutionContext> {
    private BasicCallResolutionContext(
            @NotNull BindingTrace trace,
            @NotNull LexicalScope scope,
            @NotNull Call call,
            @NotNull CangJieType expectedType,
            @NotNull DataFlowInfo dataFlowInfo,
            @NotNull ContextDependency contextDependency,
            @NotNull CheckArgumentTypesMode checkArguments,
            @NotNull ResolutionResultsCache resolutionResultsCache,
            @Nullable MutableDataFlowInfoForArguments dataFlowInfoForArguments,
            @NotNull StatementFilter statementFilter,
            boolean isAnnotationContext,
            boolean isDebuggerContext,
            boolean collectAllCandidates,
            boolean isSaveTypeInfo,
            @NotNull CallPosition callPosition,
            @NotNull Function1<CjExpression, CjExpression> expressionContextProvider,
            @NotNull LanguageVersionSettings languageVersionSettings,
            @NotNull DataFlowValueFactory dataFlowValueFactory,
            @NotNull InferenceSession inferenceSession,
            @NotNull ContextConfig config

    ) {
        super(trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments, resolutionResultsCache,
                dataFlowInfoForArguments, statementFilter, isAnnotationContext, isDebuggerContext, collectAllCandidates,
                isSaveTypeInfo, callPosition, expressionContextProvider
                ,
                languageVersionSettings,
                dataFlowValueFactory, inferenceSession, config);
    }

    @NotNull
    public static BasicCallResolutionContext create(
            @NotNull BindingTrace trace,
            @NotNull LexicalScope scope,
            @NotNull Call call,
            @NotNull CangJieType expectedType,
            @NotNull DataFlowInfo dataFlowInfo,
            @NotNull ContextDependency contextDependency,
            @NotNull CheckArgumentTypesMode checkArguments,
            boolean isAnnotationContext,
            @NotNull LanguageVersionSettings languageVersionSettings,
            @NotNull DataFlowValueFactory dataFlowValueFactory,
            @NotNull InferenceSession inferenceSession


    ) {
        return new BasicCallResolutionContext(trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
                new ResolutionResultsCacheImpl(), null,
                StatementFilter.NONE, isAnnotationContext, false, false, true,
                CallPosition.Unknown.INSTANCE, DEFAULT_EXPRESSION_CONTEXT_PROVIDER,
                languageVersionSettings,
                dataFlowValueFactory, inferenceSession, ContextConfig.DEFAULT);
    }

    @NotNull
    public static BasicCallResolutionContext create(
            @NotNull BindingTrace trace,
            @NotNull LexicalScope scope,
            @NotNull Call call,
            @NotNull CangJieType expectedType,
            @NotNull DataFlowInfo dataFlowInfo,
            @NotNull ContextDependency contextDependency,
            @NotNull CheckArgumentTypesMode checkArguments,
            boolean isAnnotationContext,
            @NotNull LanguageVersionSettings languageVersionSettings,
            @NotNull DataFlowValueFactory dataFlowValueFactory
    ) {
        return new BasicCallResolutionContext(trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
                new ResolutionResultsCacheImpl(), null,
                StatementFilter.NONE, isAnnotationContext, false, false, true,
                CallPosition.Unknown.INSTANCE, DEFAULT_EXPRESSION_CONTEXT_PROVIDER,
                languageVersionSettings,
                dataFlowValueFactory, InferenceSession.Companion.getDefault(), ContextConfig.DEFAULT);
    }

    @NotNull
    public static BasicCallResolutionContext create(
            @NotNull ResolutionContext<?> context, @NotNull Call call, @NotNull CheckArgumentTypesMode checkArguments,
            @Nullable MutableDataFlowInfoForArguments dataFlowInfoForArguments
    ) {
        return new BasicCallResolutionContext(
                context.trace, context.scope, call, context.expectedType, context.dataFlowInfo, context.contextDependency, checkArguments,
                context.resolutionResultsCache, dataFlowInfoForArguments,
                context.statementFilter, context.isAnnotationContext, context.isDebuggerContext, context.collectAllCandidates, context.isSaveTypeInfo,
                context.callPosition, context.expressionContextProvider,
                context.languageVersionSettings,
                context.dataFlowValueFactory,
                context.inferenceSession, context.config);
    }

    @NotNull
    public static BasicCallResolutionContext create(
            @NotNull ResolutionContext context, @NotNull Call call, @NotNull CheckArgumentTypesMode checkArguments
    ) {
        return create(context, call, checkArguments, null);
    }

    @Override
    protected BasicCallResolutionContext create(
            @NotNull BindingTrace trace,
            @NotNull LexicalScope scope,
            @NotNull DataFlowInfo dataFlowInfo,
            @NotNull CangJieType expectedType,
            @NotNull ContextDependency contextDependency,
            @NotNull ResolutionResultsCache resolutionResultsCache,
            @NotNull StatementFilter statementFilter,
            boolean collectAllCandidates,
            boolean isSaveTypeInfo,
            @NotNull CallPosition callPosition,
            @NotNull Function1<CjExpression, CjExpression> expressionContextProvider,
            @NotNull LanguageVersionSettings languageVersionSettings,
            @NotNull DataFlowValueFactory dataFlowValueFactory,
            @NotNull InferenceSession inferenceSession,
            @NotNull ContextConfig config
    ) {
        return new BasicCallResolutionContext(
                trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments, resolutionResultsCache,
                dataFlowInfoForArguments, statementFilter, isAnnotationContext, isDebuggerContext, collectAllCandidates,
                isSaveTypeInfo, callPosition, expressionContextProvider,
                languageVersionSettings,
                dataFlowValueFactory, inferenceSession, config);
    }

    @NotNull
    public BasicCallResolutionContext replaceCall(@NotNull Call newCall) {
        return new BasicCallResolutionContext(
                trace, scope, newCall, expectedType, dataFlowInfo, contextDependency, checkArguments, resolutionResultsCache,
                dataFlowInfoForArguments, statementFilter, isAnnotationContext, isDebuggerContext, collectAllCandidates,
                isSaveTypeInfo, callPosition, expressionContextProvider,
                languageVersionSettings,
                dataFlowValueFactory, inferenceSession, config);
    }


}

File: analysis/src/main/kotlin/org/cangnova/cangjie/resolve/calls/context/BasicCallResolutionContext.kt
/*
 * Copyright 2025 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package org.cangnova.cangjie.resolve.calls.context

import org.cangnova.cangjie.config.LanguageVersionSettings
import org.cangnova.cangjie.psi.Call
import org.cangnova.cangjie.psi.CjExpression
import org.cangnova.cangjie.resolve.StatementFilter
import org.cangnova.cangjie.resolve.binding.BindingTrace
import org.cangnova.cangjie.resolve.calls.components.InferenceSession
import org.cangnova.cangjie.resolve.calls.model.MutableDataFlowInfoForArguments
import org.cangnova.cangjie.resolve.calls.smartcasts.DataFlowInfo
import org.cangnova.cangjie.resolve.calls.smartcasts.DataFlowValueFactory
import org.cangnova.cangjie.resolve.scopes.LexicalScope
import org.cangnova.cangjie.types.CangJieType
import org.cangnova.cangjie.types.expressions.ContextConfig

/**
 * 基本调用解析上下文
 *
 * 提供调用解析所需的基本上下文信息
 */
class BasicCallResolutionContext private constructor(
    trace: BindingTrace,
    scope: LexicalScope,
    call: Call,
    expectedType: CangJieType,
    dataFlowInfo: DataFlowInfo,
    contextDependency: ContextDependency,
    checkArguments: CheckArgumentTypesMode,
    resolutionResultsCache: ResolutionResultsCache,
    dataFlowInfoForArguments: MutableDataFlowInfoForArguments?,
    statementFilter: StatementFilter,
    isAnnotationContext: Boolean,
    isDebuggerContext: Boolean,
    collectAllCandidates: Boolean,
    isSaveTypeInfo: Boolean,
    callPosition: CallPosition,
    expressionContextProvider: (CjExpression) -> CjExpression?,
    languageVersionSettings: LanguageVersionSettings,
    dataFlowValueFactory: DataFlowValueFactory,
    inferenceSession: InferenceSession,
    config: ContextConfig
) : CallResolutionContext<BasicCallResolutionContext>(
    trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
    resolutionResultsCache, dataFlowInfoForArguments, statementFilter, isAnnotationContext,
    isDebuggerContext, collectAllCandidates, isSaveTypeInfo, callPosition,
    expressionContextProvider, languageVersionSettings, dataFlowValueFactory,
    inferenceSession, config
) {

    override fun create(
        trace: BindingTrace,
        scope: LexicalScope,
        dataFlowInfo: DataFlowInfo,
        expectedType: CangJieType,
        contextDependency: ContextDependency,
        resolutionResultsCache: ResolutionResultsCache,
        statementFilter: StatementFilter,
        collectAllCandidates: Boolean,
        isSaveTypeInfo: Boolean,
        callPosition: CallPosition,
        expressionContextProvider: (CjExpression) -> CjExpression?,
        languageVersionSettings: LanguageVersionSettings,
        dataFlowValueFactory: DataFlowValueFactory,
        inferenceSession: InferenceSession,
        config: ContextConfig
    ): BasicCallResolutionContext {
        return BasicCallResolutionContext(
            trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
            resolutionResultsCache, dataFlowInfoForArguments, statementFilter, isAnnotationContext,
            isDebuggerContext, collectAllCandidates, isSaveTypeInfo, callPosition,
            expressionContextProvider, languageVersionSettings, dataFlowValueFactory,
            inferenceSession, config
        )
    }

    /**
     * 替换调用对象
     */
    fun replaceCall(newCall: Call): BasicCallResolutionContext {
        return BasicCallResolutionContext(
            trace, scope, newCall, expectedType, dataFlowInfo, contextDependency, checkArguments,
            resolutionResultsCache, dataFlowInfoForArguments, statementFilter, isAnnotationContext,
            isDebuggerContext, collectAllCandidates, isSaveTypeInfo, callPosition,
            expressionContextProvider, languageVersionSettings, dataFlowValueFactory,
            inferenceSession, config
        )
    }

    companion object {
        /**
         * 创建基本调用解析上下文
         */
        
        fun create(
            trace: BindingTrace,
            scope: LexicalScope,
            call: Call,
            expectedType: CangJieType,
            dataFlowInfo: DataFlowInfo,
            contextDependency: ContextDependency,
            checkArguments: CheckArgumentTypesMode,
            isAnnotationContext: Boolean,
            languageVersionSettings: LanguageVersionSettings,
            dataFlowValueFactory: DataFlowValueFactory,
            inferenceSession: InferenceSession
        ): BasicCallResolutionContext {
            return BasicCallResolutionContext(
                trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
                ResolutionResultsCacheImpl(), null, StatementFilter.NONE, isAnnotationContext,
                false, false, true, CallPosition.Unknown, DEFAULT_EXPRESSION_CONTEXT_PROVIDER,
                languageVersionSettings, dataFlowValueFactory, inferenceSession, ContextConfig.DEFAULT
            )
        }

        /**
         * 创建基本调用解析上下文（使用默认推断会话）
         */
        
        fun create(
            trace: BindingTrace,
            scope: LexicalScope,
            call: Call,
            expectedType: CangJieType,
            dataFlowInfo: DataFlowInfo,
            contextDependency: ContextDependency,
            checkArguments: CheckArgumentTypesMode,
            isAnnotationContext: Boolean,
            languageVersionSettings: LanguageVersionSettings,
            dataFlowValueFactory: DataFlowValueFactory
        ): BasicCallResolutionContext {
            return BasicCallResolutionContext(
                trace, scope, call, expectedType, dataFlowInfo, contextDependency, checkArguments,
                ResolutionResultsCacheImpl(), null, StatementFilter.NONE, isAnnotationContext,
                false, false, true, CallPosition.Unknown, DEFAULT_EXPRESSION_CONTEXT_PROVIDER,
                languageVersionSettings, dataFlowValueFactory, InferenceSession.default,
                ContextConfig.DEFAULT
            )
        }

        /**
         * 从已有上下文创建基本调用解析上下文
         */
        
        @JvmOverloads
        fun create(
            context: ResolutionContext<*>,
            call: Call,
            checkArguments: CheckArgumentTypesMode,
            dataFlowInfoForArguments: MutableDataFlowInfoForArguments? = null
        ): BasicCallResolutionContext {
            return BasicCallResolutionContext(
                context.trace, context.scope, call, context.expectedType, context.dataFlowInfo,
                context.contextDependency, checkArguments, context.resolutionResultsCache,
                dataFlowInfoForArguments, context.statementFilter, context.isAnnotationContext,
                context.isDebuggerContext, context.collectAllCandidates, context.isSaveTypeInfo,
                context.callPosition, context.expressionContextProvider,
                context.languageVersionSettings, context.dataFlowValueFactory,
                context.inferenceSession, context.config
            )
        }
    }
}
